File: Data Structures & Algorithms/search-2d-matrix/submission-2.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m=len(matrix)
        n=len(matrix[0])
        x=0
        y=m-1
        mid=(x+y)//2
        while x<=y:
            mid=(x+y)//2
            if matrix[mid][0]<=target and matrix[mid][n-1]>=target:
                break
            elif matrix[mid][0]>target:
                y=mid-1
            elif matrix[mid][n-1]<target:
                x=mid+1
        l=0
        r=n-1
        md=(l+r)//2
        while l<=r:
            md=(l+r)//2
            if matrix[mid][md]==target :
                return True
            elif matrix[mid][md]>target:
                r=md-1
            elif matrix[mid][md]<target:
                l=md+1
        return False

File: Data Structures & Algorithms/eating-bananas/submission-2.cpp
class Solution {
public:
    long long minEatingSpeed(vector<int>& piles, int h) {
        long long n=piles.size();
        long long mx=*max_element(piles.begin(),piles.end());
        long long mn=1;
        long long mid=(mn+mx)/2;
        long long ans=mx;
        while(mn<=mx){
            mid=(mn+mx)/2;
            long long sum=0;
            for(long long i=0;i<n;i++)
            {
                sum+=(piles[i]+mid-1)/mid;

            }if(sum<=h){
                ans=mid;
                mx=mid-1;
            }else{
                mn=mid+1;
            }
        }
        return ans;
    }
};
